cmd/kube-controller-manager/app/controllermanager.go

```diff
@@ -369,11 +369,9 @@ func StartControllers(s *options.CMServer, kubeconfig *restclient.Config, rootCl
 	// Find the list of namespaced resources via discovery that the namespace controller must manage
 	namespaceKubeClient := client("namespace-controller")
 	namespaceClientPool := dynamic.NewClientPool(restclient.AddUserAgent(kubeconfig, "namespace-controller"), restMapper, dynamic.LegacyAPIPathResolverFunc)
-	groupVersionResources, err := namespaceKubeClient.Discovery().ServerPreferredNamespacedResources()
-	if err != nil {
-		glog.Fatalf("Failed to get supported resources from server: %v", err)
-	}
-	namespaceController := namespacecontroller.NewNamespaceController(namespaceKubeClient, namespaceClientPool, groupVersionResources, s.NamespaceSyncPeriod.Duration, api.FinalizerKubernetes)
+	// TODO: consider using a list-watch + cache here rather than polling
+	gvrFn := namespaceKubeClient.Discovery().ServerPreferredNamespacedResources
```

Comment:
Can we tweak this behavior some?  For example, do 1 discovery call when constructing the function to see if TPR is supported at all on the server?  If so, always return a dynamic list of resources, and if not, return the static list and never call again?